File: 2017/js/day12.js
window.AdventOfCode.Day12 = function( input )
{
	input = input
		.split( '\n' )
		.map( ( line ) => line
			.split( ' <-> ' )[ 1 ]
			.split( ', ' )
			.map( ( id ) => parseInt( id, 10 ) )
		);

	const visited = new Map();

	const visit = ( value, id ) =>
	{
		if( visited.has( id ) )
		{
			return 0;
		}

		visited.set( id, true );

		return value.reduce( ( total, nextId ) => total + visit( input[ nextId ], nextId ), 1 );
	};

	input = input.map( visit );

	const part1 = input[ 0 ];
	const part2 = input.reduce( ( total, x ) => x > 0 ? ++total : total, 0 );

	return [ part1, part2 ];
};

File: 2025/Answers/Solutions/Day12.cs
using System;
using System.Linq;

namespace AdventOfCode;

[Answer(12)]
public class Day12 : IAnswer
{
	public Solution Solve(string input)
	{
		var part1 = 0;

		var blocks = input.Split("\n\n");
		var shapes = new int[6];

		for (var i = 0; i < 6; i++)
		{
			shapes[i] = blocks[i].Count(static c => c == '#');
		}

		foreach (var line in blocks[^1].EnumerateLines())
		{
			var i = line.IndexOf('x');
			var j = line.IndexOf(':');
			var a = line[..i].ParseInt();
			var b = line[(i + 1)..j].ParseInt();

			var ints = line[(j + 2)..].ToString().Split(' ').Select(s => s.ParseInt()).ToArray();
			var area = 0;

			for (var k = 0; k < 6; k++)
			{
				area += ints[k] * shapes[k];
			}

			if (area <= a * b)
			{
				part1++;
			}
		}

		return new(part1.ToString(), "0");
	}
}
